src/nodes/NodeConnectionManager.ts

```diff
@@ -0,0 +1,706 @@
+import type ForwardProxy from '../network/ForwardProxy';
+import type { Host, Hostname, Port } from '../network/types';
+import type { ResourceAcquire } from '../utils';
+import type { Timer } from '../types';
+import Logger from '@matrixai/logger';
+import { StartStop, ready } from '@matrixai/async-init/dist/StartStop';
+import { IdInternal } from '@matrixai/id';
+import NodeConnection from './NodeConnection';
+import * as nodesUtils from './utils';
+import * as nodesErrors from './errors';
+import { utils as validationUtils } from '../validation';
+import * as networkUtils from '../network/utils';
+import GRPCClientAgent from '../agent/GRPCClientAgent';
+import * as agentErrors from '../agent/errors';
+import * as grpcErrors from '../grpc/errors';
+import * as nodesPB from '../proto/js/polykey/v1/nodes/nodes_pb';
+import { RWLock, withF } from '../utils';
+
+type ConnectionAndLock = {
+  connection?: NodeConnection<GRPCClientAgent>;
+  timer?: NodeJS.Timer;
+  lock: RWLock;
+};
+
+/**
+ * Data structure to store all NodeConnections. If a connection to a node n does
+ * not exist, no entry for n will exist in the map. Alternatively, if a
+ * connection is currently being instantiated by some thread, an entry will
+ * exist in the map, but only with the lock (no connection object). Once a
+ * connection is instantiated, the entry in the map is updated to include the
+ * connection object.
+ */
+type NodeConnectionMap = Map<string, ConnectionAndLock>;
+
+/**
+ * This interface is provided to each `NodeConnection` in the connection map
+ * to provide necessary features. The main one being `killSelf` to allow the
+ * `NodeConnection` to tell the `NodeConnectionManager` to destroy the
+ * connection thus removing it from  the connection map.
+ */
+export interface NodeConnectionManagerInterface {
+  withConnF: typeof NodeConnectionManager.prototype.withConnF;
+  getSeedNodes: typeof NodeConnectionManager.prototype.getSeedNodes;
+  sendHolePunchMessage: typeof NodeConnectionManager.prototype.sendHolePunchMessage;
+  killSelf(): Promise<void>;
+}
+
+interface NodeConnectionManager extends StartStop {}
+@StartStop()
+class NodeConnectionManager {
+  protected nodeGraph: NodeGraph;
+  protected keyManager: KeyManager;
+  protected fwdProxy: ForwardProxy;
+  protected revProxy: ReverseProxy;
+  protected logger: Logger;
+  // Reusing terms from network domain:
+  protected connTimeoutTime: number; // - timeout for opening the connection
+  protected connConnectTime: number | undefined; // - TTL for the connection itself
+
+  // Active connections to other nodes
+  protected connections: NodeConnectionMap = new Map();
+  // Node ID -> node address mappings for the seed nodes
+  protected readonly seedNodes: NodeMapping;
+  // Max parallel connections (a.k.a. alpha constant from kademlia)
+  public readonly maxConcurrentNodeConnections: number = 3;
+
+  public constructor({
+    nodeGraph,
+    keyManager,
+    fwdProxy,
+    revProxy,
+    seedNodes = {},
```

Comment:
If the `seedNodes` is empty. There's no way to add new seed nodes into the system. Does that mean any kind of relayed connection just doesn't work? As in it will only be capable of direct connections?

Reply:
I presume so since the `NodeConnection` sends a hole punch relay message to each seed node when the connection is created. So yeah, only direct connections if there is no seed node to relay the message.

I don't think there is currently a way to add seed nodes except during the creation of the NodeConnectionManager but that is an easy change.

Reply:
That's fine for now.